fgpyo/sam/__init__.py

```diff
@@ -1231,3 +1250,285 @@ class SamOrder(enum.Enum):
+                query_sequence = primary.query_sequence
+                query_qualities = primary.query_qualities
+
+            aux = AlignedSegment(header=primary.header)
+            aux.query_name = primary.query_name
+            aux.query_sequence = query_sequence
+            aux.query_qualities = query_qualities
+
+            # Set all alignment and mapping information for this auxiliary alignment.
+            aux.cigarstring = str(hit.cigar)
+            aux.mapping_quality = 0 if hit.mapping_quality is None else hit.mapping_quality
+            aux.reference_id = primary.header.get_tid(hit.reference_name)
+            aux.reference_name = hit.reference_name
+            aux.reference_start = hit.reference_start
+            aux.is_secondary = hit.is_secondary
+            aux.is_supplementary = hit.is_supplementary
+            aux.is_proper_pair = primary.is_proper_pair if hit.is_supplementary else False
+
+            # Set all fields that relate to the template.
+            aux.is_duplicate = primary.is_duplicate
+            aux.is_forward = hit.is_forward
+            aux.is_mapped = True
+            aux.is_paired = primary.is_paired
+            aux.is_qcfail = primary.is_qcfail
+            aux.is_read1 = primary.is_read1
+            aux.is_read2 = primary.is_read2
+
+            # Set some optional, but highly recommended, SAM tags on the auxiliary alignment.
+            aux.set_tag("AS", hit.alignment_score)
+            aux.set_tag("NM", hit.edit_distance)
+            aux.set_tag("RG", primary.get_tag("RG") if primary.has_tag("RG") else None)
+            aux.set_tag("RX", primary.get_tag("RX") if primary.has_tag("RX") else None)
+
+            # Auxiliary alignment mate information points to the mate/next primary alignment.
+            aux.next_reference_id = primary.next_reference_id
+            aux.next_reference_name = primary.next_reference_name
+            aux.next_reference_start = primary.next_reference_start
+            aux.mate_is_mapped = primary.mate_is_mapped
+            aux.mate_is_reverse = primary.mate_is_reverse
+            aux.set_tag("MC", primary.get_tag("MC") if primary.has_tag("MC") else None)
+            aux.set_tag("MQ", primary.get_tag("MQ") if primary.has_tag("MQ") else None)
+            aux.set_tag("ms", primary.get_tag("ms") if primary.has_tag("ms") else None)
+
+            # Finally, set a tag that marks this alignment as a reconstituted auxiliary alignment.
+            aux.set_tag("rh", True)
+
+            yield aux
+
+    @classmethod
+    def add_all_to_template(cls, template: Template) -> Template:
+        """Rebuild a template by adding auxiliary alignments from the primary alignment tags."""
+        r1_aux = iter([]) if template.r1 is None else cls.many_pysam_from_primary(template.r1)
+        r2_aux = iter([]) if template.r2 is None else cls.many_pysam_from_primary(template.r2)
+        return Template.build(recs=chain(template.all_recs(), r1_aux, r2_aux))
```

Comment:
I think this would feel more at home on `Template`, and it's quite simple, so I don't see any harm adding e.g.:

```python
def with_all_aux_alignments() -> Template:
```

to Template.